xcm/xcm-builder/src/transactional.rs

```diff
@@ -0,0 +1,44 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use frame_support::{
+	pallet_prelude::DispatchError,
+	storage::{with_transaction, TransactionOutcome},
+};
+use xcm::latest::prelude::*;
+use xcm_executor::traits::ProcessTransaction;
+
+pub struct FrameTransactionalProcessor;
+impl ProcessTransaction for FrameTransactionalProcessor {
+	fn process_transaction<F>(f: F) -> Result<(), XcmError>
+	where
+		F: FnOnce() -> Result<(), XcmError>,
+	{
+		let transaction_outcome =
+			with_transaction(|| -> TransactionOutcome<Result<_, DispatchError>> {
+				let outcome = f();
+				match &outcome {
+					Ok(_) => TransactionOutcome::Commit(Ok(None)),
+					Err(err) => TransactionOutcome::Rollback(Ok(Some(*err))),
+				}
+			});
+		match transaction_outcome {
+			Ok(None) => Ok(()),
+			Ok(Some(err)) => Err(err),
+			Err(err) => Err(XcmError::FailedToTransactAsset(err.into())),
```

Comment:
this makes zero sense. it either success, or fail. what is the third alternative?

Reply:
The function f that we call in `process_transactional` can return an XcmError but the closure in `with_transactional` can only return an `DispatchError`. So I wrapped the `XcmError` in an Ok, so that we can pass on the `XcmError` as an result for the `process_transaction` function. with_transactional can also throw an error, so we need to match that error to be able to return an XcmError for `process_transaction`.  This solution is based on the approach in https://github.com/paritytech/substrate/blob/e374a33fe1d99d59eb24a08981090bdb4503e81b/frame/contracts/src/exec.rs#L923-L930. 

But please let me know if you know a better solution! :)

Reply:
just use #[transactional] and you won’t need this write this

Reply:
`#[transactional]` actually does almost the same thing as what was written here:

```rust
quote! {
    #(#attrs)*
    #vis #sig {
        use #crate_::storage::{with_transaction, TransactionOutcome};
        with_transaction(|| {
            let r = (|| { #block })();
            if r.is_ok() {
                TransactionOutcome::Commit(r)
            } else {
                TransactionOutcome::Rollback(r)
            }
        })
    }
}
```

However, what I think is important is to document just why exactly we'd have a `Result<Option>` type returned by `with_transaction`.